SQL_DML_HW

h6bi opened the pull request.

h6bi committed:
Added WEEK 3 HW

KseniyaLazarchik reviewed (commented):
Hi @h6bi 

Good job
Some issues


1.In WHERE condition better use UPPER/LOWER
2.I think that you use autocommit thats why NO COMMIT in your code
3.In customer update part 

Plz use in UPDATE:
WHERE customer_id = (
	SELECT c.customer_id
	FROM public.customer c
	INNER JOIN public.rental r      ON c.customer_id 	= r.customer_id 
    INNER JOIN public.payment p     	ON r.rental_id 		= p.rental_id
    GROUP BY c.customer_id
    HAVING 	COUNT(r.rental_id) 	>= 43 
		AND COUNT(p.payment_id) >= 43
    LIMIT 1
	)

4. For last part (rental and payment), you can start with (also don't forget make it reusable):

WITH insert_rent AS (
	INSERT INTO public.rental (rental_date, return_date, inventory_id, customer_id, staff_id)
	VALUES 	(
			(
			SELECT 	put your values)
			),
			(
			SELECT NOW()
			),
			(
			SELECT i.inventory_id
			FROM public.inventory i
			WHERE i.film_id = (
					SELECT f.film_id 
					FROM public.film f
					WHERE UPPER(f.title) = 'BLADE RUNNER 2049'
					LIMIT 1
					)
			),
			(
			SELECT c.customer_id	
			FROM public.customer c
			WHERE   UPPER(c.first_name)   	= 'ABEL'
			AND UPPER(c.last_name)   	= 'HORVATH'
			),
			(
			SELECT s.staff_id 	
			FROM public.staff s 
			WHERE s.store_id = (
					SELECT i.store_id
					FROM public.inventory i 
					INNER JOIN public.film f ON i.film_id = f.film_id
					WHERE UPPER(f.title) = 'BLADE RUNNER 2049'
					LIMIT 1
					) 
				LIMIT 1
				)
			)
	RETURNING 	rental_id,
			rental_date,
			return_date,
			inventory_id,
			customer_id,
			staff_id 
	)
INSERT INTO public.payment (customer_id, staff_id, rental_id, amount, payment_date)
SELECT	cte.customer_id,
	cte.staff_id,
	cte.rental_id, 
	(
	put your calculations here
	),
	NOW()
FROM insert_rent cte
RETURNING 	rental_id,
		payment_id;

1.In WHERE condition better use UPPER/LOWER
2.I think that you use autocommit thats why NO COMMIT in your codev
3.In customer update part 

Plz use in UPDATE:
WHERE customer_id = (
	SELECT c.customer_id
	FROM public.customer c
	INNER JOIN public.rental r      ON c.customer_id 	= r.customer_id 
    INNER JOIN public.payment p     	ON r.rental_id 		= p.rental_id
    GROUP BY c.customer_id
    HAVING 	COUNT(r.rental_id) 	>= 43 
		AND COUNT(p.payment_id) >= 43
    LIMIT 1
	)

4. For last part (rental and payment), you can start with (also don't forget make it reusable):

WITH insert_rent AS (
	INSERT INTO public.rental (rental_date, return_date, inventory_id, customer_id, staff_id)
	VALUES 	(
			(
			SELECT 	put your values)
			),
			(
			SELECT NOW()
			),
			(
			SELECT i.inventory_id
			FROM public.inventory i
			WHERE i.film_id = (
					SELECT f.film_id 
					FROM public.film f
					WHERE UPPER(f.title) = 'BLADE RUNNER 2049'
					LIMIT 1
					)
			),
			(
			SELECT c.customer_id	
			FROM public.customer c
			WHERE   UPPER(c.first_name)   	= 'ABEL'
			AND UPPER(c.last_name)   	= 'HORVATH'
			),
			(
			SELECT s.staff_id 	
			FROM public.staff s 
			WHERE s.store_id = (
					SELECT i.store_id
					FROM public.inventory i 
					INNER JOIN public.film f ON i.film_id = f.film_id
					WHERE UPPER(f.title) = 'BLADE RUNNER 2049'
					LIMIT 1
					) 
				LIMIT 1
				)
			)
	RETURNING 	rental_id,
			rental_date,
			return_date,
			inventory_id,
			customer_id,
			staff_id 
	)
INSERT INTO public.payment (customer_id, staff_id, rental_id, amount, payment_date)
SELECT	cte.customer_id,
	cte.staff_id,
	cte.rental_id, 
	(
	put your calculations here
	),
	NOW()
FROM insert_rent cte
RETURNING 	rental_id,
		payment_id;

h6bi committed:
DML HW correction

h6bi commented:
Hi @KseniyaLazarchik :) thank you for the feedback!

Commited the following corrections:
- added UPPER in WHERE conditions
- added COMMIT to my code
- reworked customer update and rental+payment part as requested

KseniyaLazarchik reviewed (approved):
In last part you try to use CTE (WITH ...) for multiple INSERT blocks. You can try working with an anonymous block

DO $$
DECLARE
    film_titles TEXT[] := ARRAY['BLADE RUNNER 2049', 'FORREST GUMP', 'SUPERHERO MOVIE'];
    film_title TEXT;

    v_film_id INT;
    v_inventory_id INT;
    v_store_id INT;
    v_customer_id INT;
    v_staff_id INT;
    v_rent_id INT;
    v_rental_rate NUMERIC;
    v_rental_date TIMESTAMPTZ := '2017-04-11';
    v_return_date TIMESTAMPTZ;
BEGIN
    -- Get customer once
    SELECT customer_id INTO v_customer_id
    FROM customer
    WHERE UPPER(first_name) = 'ABEL' AND UPPER(last_name) = 'HORVATH'
    ;

    FOREACH film_title IN ARRAY film_titles LOOP
        -- Get film_id
        SELECT film_id INTO v_film_id
        FROM film
        WHERE UPPER(title) = film_title
        ;

        -- Get inventory and store
        SELECT inventory_id, store_id INTO v_inventory_id, v_store_id
        FROM inventory
        WHERE film_id = v_film_id
        LIMIT 1;

        -- Get staff
        SELECT staff_id INTO v_staff_id
        FROM staff
        WHERE store_id = v_store_id
        LIMIT 1;

        -- Rental dates
        v_return_date := v_rental_date + INTERVAL '7 days';

        -- Insert into rental
        INSERT INTO rental (rental_date, inventory_id, customer_id, return_date, staff_id)
        VALUES (v_rental_date, v_inventory_id, v_customer_id, v_return_date, v_staff_id)
        RETURNING rental_id INTO v_rent_id;

        -- Get rental rate
        SELECT rental_rate INTO v_rental_rate
        FROM film
        WHERE film_id = v_film_id;

        -- Insert into payment
        INSERT INTO payment (customer_id, staff_id, rental_id, amount, payment_date)
        VALUES (
            v_customer_id,
            v_staff_id,
            v_rent_id,
            v_rental_rate * EXTRACT(DAY FROM v_return_date - v_rental_date),
            v_return_date
        );
    END LOOP;
END $$;

You can merge
